### ISSUE TITLE
Could not resolve the method in class * with ERRORLEVEL 0 

### ISSUE THREAD

**Issue by bursoft-git** [JesusFreke/smali#381]
Hi JesusFreke!

When I deodex file from MIUI firmware (Android 6.0.1):
`java -jar baksmali-2.1.2.jar -x -c boot.oat -d arm -e telephony-common.jar arm\boot.oat`
i got OUT folder and message
```
org.jf.dexlib2.analysis.AnalysisException: Could not resolve the method in class Lcom/google/android/mms/pdu/GenericPdu; at index 19
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInvokeVirtualQuick(MethodAnalyzer.java:1793)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInstruction(MethodAnalyzer.java:1023)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyze(MethodAnalyzer.java:206)
	at org.jf.dexlib2.analysis.MethodAnalyzer.<init>(MethodAnalyzer.java:141)
	at org.jf.baksmali.Adaptors.MethodDefinition.addAnalyzedInstructionMethodItems(MethodDefinition.java:463)
	at org.jf.baksmali.Adaptors.MethodDefinition.getMethodItems(MethodDefinition.java:371)
	at org.jf.baksmali.Adaptors.MethodDefinition.writeTo(MethodDefinition.java:238)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeVirtualMethods(ClassDefinition.java:327)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeTo(ClassDefinition.java:113)
	at org.jf.baksmali.baksmali.disassembleClass(baksmali.java:225)
	at org.jf.baksmali.baksmali.access$000(baksmali.java:55)
	at org.jf.baksmali.baksmali$1.call(baksmali.java:149)
	at org.jf.baksmali.baksmali$1.call(baksmali.java:147)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
opcode: invoke-virtual-quick/range
code address: 649
method: Lcom/google/android/mms/pdu/MiuiPduPersister;->persist(Lcom/google/android/mms/pdu/GenericPdu;Landroid/net/Uri;Landroid/net/Uri;JI)Landroid/net/Uri;
```
but ERRORLEVEL = 0.

Is that error or not?
This file from latest MIUI firmware http://bigota.d.miui.com/6.3.21/miui_MI3WMI4WAlpha_6.3.21_1020578f6c_6.0.zip

Thanks!

**Comment by bursoft-git** [comment]
I deodex that file with oat2dex, and compare them
There are some difference, so i think it is bug
https://drive.google.com/file/d/0B3Nba-Da-j9TbUloU2VjdU1OMGs/view?usp=sharing

**Comment by bursoft-git** [comment]
I tried to deodex this application with latest commit https://github.com/JesusFreke/smali/commit/93100e57b271cb408ec4d4247493f1336425662a , but still with no luck. Firmware not boot.

**Comment by JesusFreke** [comment]
"Firmware not boot" is not an actionable description of the problem.

**Comment by JesusFreke** [comment]
Also, sorry for the late response. This one must have slipped through the cracks :)

**Comment by JesusFreke** [comment]
With the latest commit, do you still get the "Could not resolve method..." errors when deodexing? 

(This is the type of info you should include.. think of it in terms of the info that someone looking into the problem is likely going to need to debug/troubleshoot the issue :). The more info/details you can give, the better)

**Comment by bursoft-git** [comment]
Jesus Freke, sorry for the unfortunate description of the problem.

Yes, i still get error 
```
org.jf.dexlib2.analysis.AnalysisException: Could not resolve the method in class Lcom/google/android/mms/pdu/GenericPdu; at index 19
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInvokeVirtualQuick(MethodAnalyzer.java:1835)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInstruction(MethodAnalyzer.java:1030)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyze(MethodAnalyzer.java:210)
	at org.jf.dexlib2.analysis.MethodAnalyzer.<init>(MethodAnalyzer.java:141)
	at org.jf.baksmali.Adaptors.MethodDefinition.addAnalyzedInstructionMethodItems(MethodDefinition.java:463)
	at org.jf.baksmali.Adaptors.MethodDefinition.getMethodItems(MethodDefinition.java:371)
	at org.jf.baksmali.Adaptors.MethodDefinition.writeTo(MethodDefinition.java:238)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeVirtualMethods(ClassDefinition.java:327)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeTo(ClassDefinition.java:113)
	at org.jf.baksmali.baksmali.disassembleClass(baksmali.java:225)
	at org.jf.baksmali.baksmali.access$000(baksmali.java:55)
	at org.jf.baksmali.baksmali$1.call(baksmali.java:149)
	at org.jf.baksmali.baksmali$1.call(baksmali.java:147)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
opcode: invoke-virtual-quick/range
code address: 649
method: Lcom/google/android/mms/pdu/MiuiPduPersister;->persist(Lcom/google/android/mms/pdu/GenericPdu;Landroid/net/Uri;Landroid/net/Uri;JI)Landroid/net/Uri;
```

Here is archive with the necessary files (deodexing log, boot.oat, and result of deodexing with smali and oat2dex)
https://yadi.sk/d/g4i5XE2IrVg6W

**Comment by bursoft-git** [comment]
Latest smali 2.2 WIP https://github.com/JesusFreke/smali/tree/bd349c1267ebbef2222a2985fa2ead0a95603390 also has this error
```
org.jf.dexlib2.analysis.AnalysisException: Could not resolve the method in class Lcom/google/android/mms/pdu/GenericPdu; at index 19
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInvokeVirtualQuick(MethodAnalyzer.java:1823)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInstruction(MethodAnalyzer.java:1020)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyze(MethodAnalyzer.java:200)
	at org.jf.dexlib2.analysis.MethodAnalyzer.<init>(MethodAnalyzer.java:130)
	at org.jf.baksmali.Adaptors.MethodDefinition.addAnalyzedInstructionMethodItems(MethodDefinition.java:463)
	at org.jf.baksmali.Adaptors.MethodDefinition.getMethodItems(MethodDefinition.java:371)
	at org.jf.baksmali.Adaptors.MethodDefinition.writeTo(MethodDefinition.java:238)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeVirtualMethods(ClassDefinition.java:326)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeTo(ClassDefinition.java:112)
	at org.jf.baksmali.Baksmali.disassembleClass(Baksmali.java:136)
	at org.jf.baksmali.Baksmali.access$000(Baksmali.java:43)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:60)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:58)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
opcode: invoke-virtual-quick/range
code address: 649
method: Lcom/google/android/mms/pdu/MiuiPduPersister;->persist(Lcom/google/android/mms/pdu/GenericPdu;Landroid/net/Uri;Landroid/net/Uri;JI)Landroid/net/Uri;
```

**Comment by JesusFreke** [comment]
Oh, I see what the problem here is. yech.

    move-object/from16 v0, p1
    instance-of v2, v0, Lcom/google/android/mms/pdu/MultimediaMessagePdu;
    if-eqz v2, :cond_2bb
    nop
    nop
    invoke-virtual-quick/range {p1 .. p1}, vtable@19

The problem is that it has to move the p1 value down to a lower register before it uses instance-of on it. But then, if the instance-of succeeded, it goes back to using p1 to call a method on it.

I have some special case logic to narrow the register type fur an instance-of followed by a conditional, but it doesn't handle aliased registers like that.

**Comment by JesusFreke** [comment]
This should be fixed in the v2.2_WIP branch now

**Comment by bursoft-git** [comment]
Commit https://github.com/JesusFreke/smali/commit/db49ae1d03e918a21c2ce7afe49a5ce841d950a8 broke deodexing file apache-xml.jar, which deodexing fine with commit https://github.com/JesusFreke/smali/commit/8a5a6e3fc57adffebd9c9b717af85b66159a1208

```
java -jar baksmali-2.2.jar x -a 17 -d _framework "_framework\apache-xml.odex" 

org.jf.dexlib2.analysis.AnalysisException: Could not resolve the method in class Ljava/lang/Throwable; at index 27
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInvokeVirtualQuick(MethodAnalyzer.java:1844)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInstruction(MethodAnalyzer.java:1016)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyze(MethodAnalyzer.java:200)
	at org.jf.dexlib2.analysis.MethodAnalyzer.<init>(MethodAnalyzer.java:130)
	at org.jf.baksmali.Adaptors.MethodDefinition.addAnalyzedInstructionMethodItems(MethodDefinition.java:463)
	at org.jf.baksmali.Adaptors.MethodDefinition.getMethodItems(MethodDefinition.java:371)
	at org.jf.baksmali.Adaptors.MethodDefinition.writeTo(MethodDefinition.java:238)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeDirectMethods(ClassDefinition.java:281)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeTo(ClassDefinition.java:111)
	at org.jf.baksmali.Baksmali.disassembleClass(Baksmali.java:136)
	at org.jf.baksmali.Baksmali.access$000(Baksmali.java:43)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:60)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:58)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
opcode: invoke-virtual-quick
code address: 26
method: Lorg/apache/xml/utils/DefaultErrorHandler;->ensureLocationSet(Ljavax/xml/transform/TransformerException;)V
```
[apache-xml.zip](https://github.com/JesusFreke/smali/files/290476/apache-xml.zip)
